internal/transport/http2_server.go

```diff
@@ -347,14 +416,14 @@ func (t *http2Server) operateHeaders(frame *http2.MetaHeadersFrame, handle func(
 	}
 	s.ctx = peer.NewContext(s.ctx, pr)
 	// Attach the received metadata to the context.
-	if len(state.data.mdata) > 0 {
-		s.ctx = metadata.NewIncomingContext(s.ctx, state.data.mdata)
+	if len(mdata) > 0 {
+		s.ctx = metadata.NewIncomingContext(s.ctx, mdata)
 	}
-	if state.data.statsTags != nil {
-		s.ctx = stats.SetIncomingTags(s.ctx, state.data.statsTags)
+	if statsTags, ok := mdata["grpc-tags-bin"]; ok && len(statsTags) > 0 {
```

Comment:
Also a _very_ minor optimization, these can be moved inside `if len(mdata) > 0`

Reply:
Done!